### PR TITLE
AA-412: Separate 'nonceKey' and 'nonceSequence' values

### PR DESCRIPTION
**Pull request by forshtat**

### CODE REVIEW

File: RIPS/rip-7712.md [hunk 1]
```diff
@@ -39,26 +39,43 @@ with ERC-4337, as all ERC-4337 Smart Contract Accounts already use the solution
...
+### Constants
+
+| Name                     | Value |
+|--------------------------|-------|
+| AA_BASE_GAS_COST         | 15000 |
+| AA_BASE_GAS_COST_RIP7712 | 10000 |
+
 ### Non-sequential nonce support
 
-We propose an introduction of a separate 2-dimensional `nonce` value used when contracts initiate a transaction.
```
Comment (drortirosh) [inline RIPS/rip-7712.md:44]: This two-dimensional nonce mechanism is exposed to the EVM in a `NonceManager` pre-deployed contract
located at the `AA_NONCE_MANAGER` address.

We introduce a new `nonceKey` field in the AA_TX_TYPE transaction.

If nonceKey is zero (non-existent), then the old system nonce value is incremented, and the transaction's base cost is `AA_BASE_GAS_COST`.

If nonceKey is non-zero, then the nonce manager is in use for validation, the `nonceSeq` is incremented, and the base cost is `AA_BASE_GAS_COST_RIP7712`, as the validation gas cost will include the cost of calling the nonce manager. In this case, The old system nonce is not in use and not incremented.

During account deployment transaction, the nonceKey MUST be zero.
The old system  nonce is not incremented before the transaction: the CREATE/CREATE2 opcode used by the deployer will cause the "nonce bump"

File: RIPS/rip-7712.md [hunk 2]
```diff
@@ -39,26 +39,42 @@ with ERC-4337, as all ERC-4337 Smart Contract Accounts already use the solution
...
-The `nonce` is [validated and incremented](#nonce-validation-frame) on-chain before the rest of the validation code.
+The two-dimensional nonce is [validated and incremented](#nonce-validation-frame) on-chain
+as part of a `AA_TX_TYPE` transaction validation before the rest of the validation code.
 
 ### The old `nonce` account parameter use
 
-The old 64-bit `nonce` account parameter is used when the `key` of the `AA_TX_TYPE` transaction's `nonce` is set to 0.
-Notice that this also prevents the previously signed transactions from becoming valid again.
+The old 64-bit `nonce` account parameter is used when the `nonceKey` of the `AA_TX_TYPE` transaction is set to 0.
+Notice that this also prevents the previously signed `AA_TX_TYPE` transactions from becoming valid again.
```
Comment (drortirosh) [inline RIPS/rip-7712.md]: this notice is useless

File: RIPS/rip-7712.md [hunk 3]
```diff
@@ -39,36 +39,47 @@ with ERC-4337, as all ERC-4337 Smart Contract Accounts already use the solution
...
-The `nonce` is [validated and incremented](#nonce-validation-frame) on-chain before the rest of the validation code.
+The two-dimensional nonce is [validated and incremented](#nonce-validation-frame) on-chain
+as part of a `AA_TX_TYPE` transaction validation before the rest of the validation code.
 
-### The old `nonce` account parameter use
+### Nonce validation
 
-The old 64-bit `nonce` account parameter is used when the `key` of the `AA_TX_TYPE` transaction's `nonce` is set to 0.
-Notice that this also prevents the previously signed transactions from becoming valid again.
+The first validation on a transaction is of its nonce.
```
Comment (shahafn) [inline RIPS/rip-7712.md]:
```suggestion
```

Comment (forshtat) [reply RIPS/rip-7712.md]: Removed this statement, replaced with something more legible.

**Review by drortirosh — APPROVED** [review]

File: RIPS/rip-7712.md [hunk 4]
```diff
@@ -39,48 +39,62 @@ with ERC-4337, as all ERC-4337 Smart Contract Accounts already use the solution
...
-During deployment, `nonceKey` must be 0, and the old `nonce` parameter is only incremented **once** during the entire transaction.
-such that after the successful deployment `AA_TX_TYPE` transaction the `sender` will have an old `nonce` value of 1.
+### Deployment transaction
 
+During the account deployment transaction, the `nonceKey` MUST be zero.
+The old system `nonce` parameter is not incremented before the transaction:
+the `CREATE` or `CREATE2` opcode used by the deployer
+will cause the nonce to be incremented to 1 according to [EIP-161](https://eips.ethereum.org/EIPS/eip-161).
 
 #### Nonce validation frame
```
Comment (shahafn) [inline RIPS/rip-7712.md:87]:
```suggestion
#### RIP-7560 Validation phase changes
```

File: RIPS/rip-7712.md [hunk 5]
```diff
@@ -39,48 +39,62 @@ with ERC-4337, as all ERC-4337 Smart Contract Accounts already use the solution
...
-This `nonce` is exposed to the EVM in a `NonceManager` pre-deployed contract located at the `AA_NONCE_MANAGER` address.
+This two-dimensional nonce mechanism is exposed to the EVM in a `NonceManager` pre-deployed contract
+located at the `AA_NONCE_MANAGER` address.
 
-The `nonce` is [validated and incremented](#nonce-validation-frame) on-chain before the rest of the validation code.
+The two-dimensional nonce is [validated and incremented](#nonce-validation-frame) on-chain
+as part of a `AA_TX_TYPE` transaction validation before the rest of the validation code.
 
-### The old `nonce` account parameter use
+### Nonce validation
```
Comment (shahafn) [inline RIPS/rip-7712.md:67]:
```suggestion
### Nonce validation frame
```

File: RIPS/rip-7712.md [hunk 6]
```diff
@@ -39,48 +39,62 @@ with ERC-4337, as all ERC-4337 Smart Contract Accounts already use the solution
...
 
-The old 64-bit `nonce` account parameter is used when the `key` of the `AA_TX_TYPE` transaction's `nonce` is set to 0.
-Notice that this also prevents the previously signed transactions from becoming valid again.
+We propose to introduce a new validation frame to the RIP-7560 transaction validation flow.
+This validation frame performs an on-chain nonce validation and runs before all other frames in a transaction.
```
Comment (shahafn) [inline RIPS/rip-7712.md:70]:
```suggestion
This validation frame performs an on-chain nonce validation (a call to `NonceManager`) and runs before all other frames in a transaction.
The `NonceManager` is invoked with the following data:

```
sender{20 bytes} nonceKey{24 bytes} nonceSeq{8 bytes}
```

```

File: RIPS/rip-7712.md [hunk 7]
```diff
@@ -39,48 +39,62 @@ with ERC-4337, as all ERC-4337 Smart Contract Accounts already use the solution
...
 
-The old `nonce` account parameter remains in use for transactions initiated by EOAs and for the `CREATE` opcode.
-It is not incremented when `sender` makes the `AA_TX_TYPE` transaction.
+If nonceKey is zero (non-existent), the nonce field is compared to the system nonce,
+and then the system nonce is incremented.
+The transaction's base cost is `A_BASE_GAS_COST`.
```
Comment (shahafn) [inline RIPS/rip-7712.md:74]:
```suggestion
The transaction's base cost is `A_BASE_GAS_COST`.

The gas costs of this validation frame are counted towards the total for a transaction,
and is deducted from the account's `validationGasLimit`.
```

File: RIPS/rip-7712.md [hunk 8]
```diff
@@ -39,48 +39,62 @@ with ERC-4337, as all ERC-4337 Smart Contract Accounts already use the solution
...
-During deployment, `nonceKey` must be 0, and the old `nonce` parameter is only incremented **once** during the entire transaction.
-such that after the successful deployment `AA_TX_TYPE` transaction the `sender` will have an old `nonce` value of 1.
+### Deployment transaction
 
+During the account deployment transaction, the `nonceKey` MUST be zero.
+The old system `nonce` parameter is not incremented before the transaction:
+the `CREATE` or `CREATE2` opcode used by the deployer
+will cause the nonce to be incremented to 1 according to [EIP-161](https://eips.ethereum.org/EIPS/eip-161).
 
 #### Nonce validation frame
```
Comment (shahafn) [inline RIPS/rip-7712.md:88]:
```suggestion

```

File: RIPS/rip-7712.md [hunk 9]
```diff
@@ -39,48 +39,62 @@ with ERC-4337, as all ERC-4337 Smart Contract Accounts already use the solution
...
 
-Before the first validation frame is applied during the validation phase,
+Before the first RIP-7560 transaction type defined validation frame is applied during the validation phase,
 the `NonceManager` is invoked with the following data:
 
 ```
 sender{20 bytes} nonceKey{24 bytes} nonceSeq{8 bytes}
 ```
```
Comment (shahafn) [inline RIPS/rip-7712.md:95]:
```suggestion
```

File: RIPS/rip-7712.md [hunk 10]
```diff
@@ -39,48 +39,62 @@ with ERC-4337, as all ERC-4337 Smart Contract Accounts already use the solution
...
 
-The gas costs of this execution frame are counted towards the total for a transaction.
+The gas costs of this validation frame are counted towards the total for a transaction,
+and is deducted from the account's `validationGasLimit`.
```
Comment (shahafn) [inline RIPS/rip-7712.md:98]:
```suggestion
```

**Pull request merged**